### ISSUE TITLE
[Build] `re2` not compiled when statically building with `onnxruntime_BUILD_UNIT_TESTS=OFF`

### ISSUE THREAD

**Issue by cyanic-selkie** [microsoft/onnxruntime#22513]
### Describe the issue

I have compiled `onnxruntime` with the following command (on linux):

`./build.sh --cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF --config Release --parallel --compile_no_warning_as_error --skip_submodule_sync --update --build --build_dir build/linux`

When linking this to a rust library (via [ort](https://github.com/pykeio/ort)), there seems to be an [issue](https://github.com/pykeio/ort/issues/304) where, if the flag `--cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF` is given, `re2` isn't compiled (only when building a static library, apparently, it gets built when building the shared library).

### Target platform

Linux x86_64

### Build script

./build.sh --cmake_extra_defines=onnxruntime_BUILD_UNIT_TESTS=OFF --config Release --parallel --compile_no_warning_as_error --skip_submodule_sync --update --build --build_dir build/linux

### Error / output

There is no error at build time, linking downstream fails as described above.

**Comment by snnn** [comment]
I just checked https://github.com/microsoft/onnxruntime/blob/main/cmake/external/onnxruntime_external_deps.cmake, re2 doesn't seem to be optional. It should always get built.  Sorry I have no idea how it would happen.

**Comment by snnn** [comment]
Do you have more logs?  Can you set FETCHCONTENT_TRY_FIND_PACKAGE_MODE to NEVER?

**Comment by snnn** [comment]
I tried the command on Linux(sorry I don't have a Mac machine). It did not have any problem. I noticed that you didn't generate a dynamic onnxruntime lib.  So, your application needs to link to all onnxruntime's internal libraries and their dependencies(including re2). That's a lot. 
And since you disabled all executable binaries and all *.dylib targets, there is no link stage. Nothing needs be linked. Therefore re2 even won't get built. This is expected.
